src/exchanges/populate.ts

```diff
@@ -0,0 +1,331 @@
+import { pipe, tap, map } from 'wonka';
+import { Exchange, Operation } from '../types';
+import {
+  DocumentNode,
+  buildClientSchema,
+  visitWithTypeInfo,
+  TypeInfo,
+  FragmentDefinitionNode,
+  SelectionSetNode,
+  GraphQLSchema,
+  IntrospectionQuery,
+  FragmentSpreadNode,
+  GraphQLObjectType,
+  GraphQLInterfaceType,
+  ASTNode,
+} from 'graphql';
+import { visit } from 'graphql';
+
+interface ExchangeArgs {
+  schema: IntrospectionQuery;
+}
+
+/** An exchange for auto-populating mutations with a required response body. */
+export const populateExchange = ({
+  schema: ogSchema,
+}: ExchangeArgs): Exchange => ({ forward }) => {
+  const schema = buildClientSchema(ogSchema);
+  /** List of operation keys that have already been parsed. */
+  const parsedOperations = new Set<number>();
+  /** List of operation keys that have not been torn down. */
+  const activeOperations = new Set<number>();
+  /** Collection of fragments used by the user. */
+  let userFragments: UserFragmentMap = {};
+  /** Collection of type fragments. */
+  let typeFragments: TypeFragmentMap = {};
+
+  /** Handle mutation and inject selections + fragments. */
+  const handleIncomingMutation = (op: Operation) => {
+    if (op.operationName !== 'mutation') {
+      return op;
+    }
+
+    const activeSelections = Object.entries(typeFragments).reduce(
+      (state, [key, value]) => ({
+        ...state,
+        [key]: value.filter(s => activeOperations.has(s.key)),
+      }),
+      typeFragments
+    );
+
+    return {
+      ...op,
+      query: addFragmentsToQuery({
+        schema,
+        typeFragments: activeSelections,
+        userFragments: userFragments,
+        query: op.query,
+      }),
+    };
+  };
+
+  /** Handle query and extract fragments. */
+  const handleIncomingQuery = ({ key, operationName, query }: Operation) => {
+    activeOperations.add(key);
+
+    if (operationName !== 'query' || parsedOperations.has(key)) {
+      return;
+    }
+
+    parsedOperations.add(key);
+
+    const {
+      fragments: newFragments,
+      selections: newSelections,
+    } = extractSelectionsFromQuery({
+      schema,
+      query,
+    });
+
+    userFragments = newFragments.reduce(
+      (state, fragment) => ({
+        ...state,
+        [fragment.name.value]: fragment,
+      }),
+      userFragments
+    );
+
+    typeFragments = newSelections.reduce((state, { selections, type }) => {
+      const current = state[type] || [];
+      const entry: TypeFragment = {
```

Comment:
Code Golfing question cc @JoviDeCroock 

We may be able to reduce the size here by using the GraphQL constructors (i.e. `new NameNode`, etc). They're probably already included so we may as well try to use them and see whether it reduces the size here a little

Reply:
This should be the case yes, should also enforce some more streamlined objects shapes which v8 loves

Reply:
Ooft this would be nice. I tried this at first but got errors that XyzNode is a type.

Will play around.